File: architecture-skills/object-oriented-programming-paradigm/examples/object-relationships/object-relationships.js
// Object relationships in JavaScript.
//
// Every reference looks alike; signal composition by creating parts inside
// the class and keeping the field private (#courses). Aggregation accepts
// objects from outside. Association passes collaborators as method parameters.

class Book {
  constructor(title) {
    this.title = title;
  }
}

class Library {
  constructor(books = []) {
    this.books = books;
  }

  lookup(title) {
    return this.books.find((b) => b.title === title) ?? null;
  }
}

class Professor {
  constructor(name) {
    this.name = name;
  }

  // Association: library is a parameter — no stored reference.
  lookupBook(library, title) {
    const book = library.lookup(title);
    if (!book) {
      return `${this.name}: '${title}' not in catalog`;
    }
    return `${this.name} looked up '${book.title}' in the library`;
  }
}

class Course {
  constructor(title) {
    this.title = title;
  }
}

class Department {
  #courses = []; // composition: private, created and owned here

  constructor(name, professors) {
    this.name = name;
    this.professors = [...professors]; // aggregation: references from outside
  }

  addCourse(title) {
    this.#courses.push(new Course(title));
  }

  courseTitles() {
    return this.#courses.map((c) => c.title);
  }

  professorNames() {
    return this.professors.map((p) => p.name);
  }

  dissolve() {
    this.professors = []; // drop references — never destroy professors
    this.#courses = [];   // composed parts die with the whole
  }

  printRoster() {
    console.log(`${this.name} roster: ${this.professorNames().join(", ")}`);
  }

  printCourses() {
    const titles = this.courseTitles();
    console.log(
      titles.length === 0
        ? `${this.name} courses: (none)`
        : `${this.name} courses: ${titles.join(", ")}`
    );
  }
}

const library = new Library([new Book("Algorithms"), new Book("Calculus")]);

const alice = new Professor("Dr. Alice");
const bob = new Professor("Dr. Bob");

const cs = new Department("Computer Science", [alice, bob]);
cs.addCourse("Algorithms");
cs.addCourse("Data Structures");

const math = new Department("Mathematics", [alice]);
math.addCourse("Calculus");

console.log("=== Association (uses-a) ===");
console.log(alice.lookupBook(library, "Algorithms"));

console.log("\n=== Aggregation (has-a) ===");
cs.printRoster();
math.printRoster();
cs.dissolve();
console.log(`After dissolving CS — Alice still exists: ${alice.name}`);
math.printRoster();

console.log("\n=== Composition (owns-a) ===");
math.printCourses();
math.dissolve();
math.printCourses();

File: architecture-skills/object-oriented-programming-paradigm/examples/object-relationships/object-relationships.java
// Object relationships in Java.
//
// The GC erases the memory-level distinction — every field is a reference.
// Composition means the whole constructs the part and cancels it in close();
// aggregation means the part arrives via constructor and is left alone at
// teardown. Java identifiers cannot contain hyphens, so the launcher class
// is named after the concept instead of the file.

import java.util.ArrayList;
import java.util.List;

class ObjectRelationshipsDemo {

    record Book(String title) {}

    static final class Library {
        private final List<Book> books = new ArrayList<>();

        Library(List<Book> initial) {
            books.addAll(initial);
        }

        Book lookup(String title) {
            return books.stream()
                    .filter(b -> b.title().equals(title))
                    .findFirst()
                    .orElse(null);
        }
    }

    static final class Professor {
        private final String name;

        Professor(String name) {
            this.name = name;
        }

        String name() {
            return name;
        }

        // Association: library is a method parameter, not a stored field.
        String lookupBook(Library library, String title) {
            Book book = library.lookup(title);
            if (book == null) {
                return name + ": '" + title + "' not in catalog";
            }
            return name + " looked up '" + book.title() + "' in the library";
        }
    }

    record Course(String title) {}

    static final class Department {
        private final String name;
        private final List<Professor> professors; // aggregation: injected
        private final List<Course> courses = new ArrayList<>(); // composition

        Department(String name, List<Professor> professors) {
            this.name = name;
            this.professors = new ArrayList<>(professors);
        }

        void addCourse(String title) {
            courses.add(new Course(title)); // whole creates the part
        }

        List<String> professorNames() {
            return professors.stream().map(Professor::name).toList();
        }

        List<String> courseTitles() {
            return courses.stream().map(Course::title).toList();
        }

        void dissolve() {
            professors.clear(); // drop references — never destroy professors
            courses.clear();    // composed parts cancelled with the whole
        }

        void printRoster() {
            System.out.print(name + " roster:");
            for (Professor p : professors) {
                System.out.print(" " + p.name());
            }
            System.out.println();
        }

        void printCourses() {
            if (courses.isEmpty()) {
                System.out.println(name + " courses: (none)");
                return;
            }
            System.out.print(name + " courses:");
            for (Course c : courses) {
                System.out.print(" " + c.title());
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        Library library = new Library(List.of(
                new Book("Algorithms"),
                new Book("Calculus")
        ));

        Professor alice = new Professor("Dr. Alice");
        Professor bob = new Professor("Dr. Bob");

        Department cs = new Department("Computer Science", List.of(alice, bob));
        cs.addCourse("Algorithms");
        cs.addCourse("Data Structures");

        Department math = new Department("Mathematics", List.of(alice));
        math.addCourse("Calculus");

        System.out.println("=== Association (uses-a) ===");
        System.out.println(alice.lookupBook(library, "Algorithms"));

        System.out.println("\n=== Aggregation (has-a) ===");
        cs.printRoster();
        math.printRoster();
        cs.dissolve();
        System.out.println("After dissolving CS — Alice still exists: " + alice.name());
        math.printRoster();

        System.out.println("\n=== Composition (owns-a) ===");
        math.printCourses();
        math.dissolve();
        math.printCourses();
    }
}
